File: recursion/factorial.cpp
#include <iostream>

int factorial(int n)
{
    if (n==1 || n==0){
        return 1;
    }
    return n*factorial(n-1);
}

int factorialTailRecursion(int n, int result=1)
{
    if (n==1 || n==0)
        return result;
    return factorialTailRecursion(n-1, result*n);
}

int main()
{
    std::cout << factorialTailRecursion(4);
}

File: recursion/factorial.py
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n-1)

def factorialTailRecursion(n, k=1):
    if n == 0:
        return k
    return factorialTailRecursion(n-1, n*k)

print(factorial(5))
print(factorialTailRecursion(5))